proposals/ytt/004-schema-validation/README.md

```diff
@@ -0,0 +1,1523 @@
+  - [... when the data value is a type that has length](#required-input-for-types-with-length)
+  - [... when the data value is a type that has an empty value](#required-input-for-types-with-empty-value)
+  - [... for all other cases](#required-input-for-types-with-no-empty-value)
+- [Require value to be from a defined set (Enumeration)](#use-case-enumeration)
+- [Sections of configuration are mutually exclusive (Union Structures)](#use-case-union-structures)
+- [Validations involving multiple Data Values (Multiple Nodes)](#use-case-multiple-nodes)
+- [Validate nullable Data Values](#use-case-validations-over-nullables)
+- [Apply validations in specific situations (Conditional Validations)](#use-case-conditional-validations)
+- [Disable all validations in an "emergency" (Disable Validations)](#use-case-disable-validations)
+- [Reuse a set of validations over a commonly-shaped set of Data Values (Programatic Validations)](#use-case-programmatic-validations)
+- [Export Schema in OpenAPI v3 format](#use-case-exporting-schema-in-openapi-format)
+- [Validate an output document](#use-case-validating-an-output-document)
+
+The following sections detail each use case, in turn...
+
+#### Use Case: Required input
+
+When the author cannot/will not provide a default value for a data value and requires the consumer to provide it.
+
+Where possible, authors are encouraged to use a type-specific "non-empty" validation. As a fallback, they may opt for nullable+not_null.
+
+##### Required input for types with length
+
+**String Values**
+
+The recommended way for an author to indicate required string input is to expect a non-empty string:
+
+```yaml
+#@data/values-schema
+---
+dex:
+  #@assert/validate min_len=1
+  namespace: ""
+  #@assert/validate min_len=1
+  username: ""
+```
+
+This case is so common (it is our understanding that most string-based data values are meant to be non-empty), we offer this convenience:
+
+```yaml
+#@data/values-schema
+#@assert/validate-string-defaults min_len=1
+---
+dex:
+  namespace: ""
+  username: ""
+```
+
+Which is equivalent to the prior example, but applies the default to any descendent that contains a string-typed value. If a particular node already has an `@assert/validate` annotation, it is skipped.
+
+**Array Values**
+
+Similarly, a non-empty value of an array has at least one element:
+
+```yaml
+#@data/values-schema
+---
+dex:
+  oauth2:
+    #@assert/validate min_len=1
+    responseTypes:
+    - ""
+```
+
+Array-typed values are less common. And when present are not necessarily _as_ common to require a non-empty value as string-typed values. So, there is no corresponding shorthand annotation for requiring a non-empty array.
+
+
+##### Required input for types with empty value
+
+For some integer-typed (or float-typed) data values, a zero value may be invalid and thus can represent "empty".
+
+```yaml
+#@data/values-schema
+dex:
+  #@assert/validate min=1024, max=65535
```

Comment:
do we or will we provide something like MAX_INT ? I am wondering about the case where I have a parser that only fits signed32 bit integers for example, and if they are above that I want it to be invalid at ytt time before my post-processing parser falls over

Reply:
Starlark support 32-bit integers, but are stored in upwards of 64-bits to avoid overflows. So, if I understand correctly, I think your underlying concern might be addressed.

Practically speaking, we've yet to hear about someone needing the use the full range of integers. 🤔